// express/ExpressNode.js
const nodeModule = require('../nodeModule');
const BackEndRoutes = require('../express/BackEndRoute');
class ExpressNode extends nodeModule{
    constructor(){
        'use strict';
        super();
        this.path = super.__path(); //get require
        //--express
        this.express =  super.__express(); //get require
        this.app = this.express();

        //--mysql
        this._mysql = super.__mysql(); //get require

        //body-parser
        this._body_parser = super.__body_parser();

        //-- CONSTANTS.js
        this._constants = super.getconstantList();

        //--routes for backend
        this._back_end_routes = new BackEndRoutes(this._mysql, this.express, this.app, this._constants, this._body_parser);
        
        //--routes for frontend
        this._front_end_routes = null;
    }

    setEJSViewEngine(){
        this.app.set('views', this.path.join(__dirname, '../../../views'));
        this.app.set('view engine', 'ejs');
    }

    setport(){
        this.app.listen(this._constants.PORT, () => console.log("server is running on port " + this._constants.PORT) );
    }

    setHTMLPath(){
        this.app.get('/', function(req, res) {
            res.render('index');
        });
    }

    initExpress(){
        this.app.use(this.express.static(this.path.join(__dirname, '../../../../public')));
        this.setEJSViewEngine();
        this.setHTMLPath();
        this.setport();
    }
}

module.exports = ExpressNode;
