// express/queries.js
module.exports = {
    countAllProjects: 'select count(*) as num_of_project from projects',
    project:'select * from projects where project_id= ',
    project_images: 'select img.link, img.credit '+
                    'from projects as p , project_images as p_img, images as img '+
                    'where p.project_id = p_img.project_id and '+
                    'p_img.image_id = img.image_id and p.project_id = ',
    project_skills: 'select sk.skill_name, sk.skill_color '+
                    'from projects as p , project_skills as p_sk, skills as sk ' +
                    'where p.project_id = p_sk.project_id and '+
                    'p_sk.skill_id = sk.skill_id and p.project_id = ',
    project_videos: 'select vid.link '+
                    'from projects as p , project_videos as p_vid, videos as vid '+
                    'where p.project_id = p_vid.project_id and '+
                    'p_vid.video_id = vid.video_id and p.project_id = ',
    project_documentations: 'select doc.doc_title, doc.link '+
                            'from projects as p , project_documentations as p_doc, documentations as doc '+
                            'where p.project_id = p_doc.project_id and '+
                            'p_doc.doc_id = doc.doc_id and p.project_id = ',
    resume: 'select full_name, introduction, objective from resume ',
    education: 'select e.school_name, e.degree, e.start_term, e.end_term, e.GPA, img.link, img.credit '+
               'from educations as e, images as img '+
               'where e.education_icon_id = img.image_id '+
               'order by e.school_id desc ',
    experiences: 'select exp.exp_title, exp.start, exp.end, exp.exp_description, img.link, img.credit '+
                 'from experiences as exp, images as img '+
                 'where exp.exp_icon_id = img.image_id '+
                 'order by exp.exp_id desc ',
    achievements: 'select achievement_title, year, achievement_description, link, credit '+
                  'from achievement as a, images as img '+
                  'where a.achievement_icon_id = img.image_id '+
                  'order by year desc ',
    languages: 'select language, fluency_point, link, credit '+
               'from languages as l, images as img '+
               'where l.lang_icon_id = img.image_id ',
    resume_skills: 'select skill_name, skill_color, skill_point, link, credit '+
                   'from resume, resume_skills as rs, skills as s, images as img '+
                   'where resume.resume_id = rs.resume_id and '+
                   'rs.skill_id = s.skill_id and '+
                   's.skill_icon_id = img.image_id ',
    projectsAll: 'select name, description, start, end from projects ',
    contacts: 'select contact_type, contact_content, link, credit '+
              'from contacts, images as img '+
              'where contacts.contact_icon_id = img.image_id ',
    resume_project_skills: 'select p.name, p.project_id, sk.skill_name '+
                            'from projects as p , project_skills as p_sk, skills as sk '+
                            'where p.project_id = p_sk.project_id and '+
                            'p_sk.skill_id = sk.skill_id and p.project_id = ',
};
